### PR TITLE
Add support for sdsdot

### PR DESCRIPTION
**Pull request by fabiomestre**

### CODE REVIEW

File: src/interface/blas1_interface.hpp [hunk 1]
```diff
@@ -322,7 +360,43 @@ typename ValueType<container_0_t>::type _dot(executor_t &ex, index_t _N,
...
+ */
+template <typename executor_t, typename container_0_t, typename container_1_t,
+        typename index_t, typename increment_t>
+typename ValueType<container_0_t>::type _sdsdot(executor_t &ex, index_t _N, float sb,
+                                             container_0_t _vx,
+                                             increment_t _incx,
+                                             container_1_t _vy,
+                                             increment_t _incy) {
+    using element_t = typename ValueType<container_0_t>::type;
+    auto res = std::vector<element_t>(1);
```
Comment (hjabird) [inline src/interface/blas1_interface.hpp:393]:
```
element_t res{};
```
I don't think we need std::vector.

File: include/interface/blas1_interface.h [hunk 2]
```diff
@@ -76,6 +76,33 @@ template <typename executor_t, typename container_0_t, typename container_1_t,
 typename executor_t::policy_t::event_t _dot(
     executor_t &ex, index_t _N, container_0_t _vx, increment_t _incx,
     container_1_t _vy, increment_t _incy, container_2_t _rs);
+
+/**
+ * \brief Computes the inner product of two vectors with extended precision
```
Comment (hjabird) [inline include/interface/blas1_interface.h:81]: Does extended precision just mean double precision?

File: include/interface/blas1_interface.h [hunk 3]
```diff
@@ -76,6 +76,33 @@ template <typename executor_t, typename container_0_t, typename container_1_t,
...
+ * @tparam container_0_t Buffer Iterator
+ * @tparam container_1_t Buffer Iterator
+ * @tparam container_2_t Buffer Iterator
+ * @tparam index_t Index type
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes. If size 0, the result will be sb.
+ * @param sb scalar to add to the results of the inner product.
+ * @param _vx Buffer holding input vector x
+ * @param _incx Increment in x axis
```
Comment (hjabird) [inline include/interface/blas1_interface.h:94]: I think it is better to describe this as a stride. There are no x and y axes.

File: test/unittest/blas1/blas1_sdsdot_test.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+/**
+ * Operator overload to print the test name properly
+ */
+std::ostream& operator<<(std::ostream& os, const api_type& type) {
+  if (type == api_type::event) {
+    return os << "event";
+  } else {
+    return os << "result";
+  }
+}
```
Comment (hjabird) [inline test/unittest/blas1/blas1_sdsdot_test.cpp:41]: Should be part of the common code? There is a specialization of the code for dumping floats, double and and half somewhere.

File: test/unittest/blas1/blas1_sdsdot_test.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+template <typename scalar_t>
+using combination_t = std::tuple<api_type, int, float, int, int>;
```
Comment (hjabird) [inline test/unittest/blas1/blas1_sdsdot_test.cpp:44]:
```
template <typename scalar_t>
using combination_t = std::tuple<api_type, int, scalar_t, int, int>;
```
?

File: test/unittest/blas1/blas1_sdsdot_test.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+  index_t incY;
+  api_type api;
+  std::tie(api, N, sb, incX, incY) = combi;
+
+  using data_t = utils::data_storage_t<scalar_t>;
+
+  /* Sycl Buffers do not work with size = 0. So setting input vectors to size
+   * one to test the edge case where if size equals 0 then sb should be
+   * returned. */
+  index_t vectorSize = N > 0 ? N : 1;
```
Comment (hjabird) [inline test/unittest/blas1/blas1_sdsdot_test.cpp:63]: We claim that `N` can be zero in the docs. This then says that this doesn't work?

File: src/interface/blas1_interface.hpp [hunk 7]
```diff
@@ -322,7 +360,43 @@ typename ValueType<container_0_t>::type _dot(executor_t &ex, index_t _N,
   auto gpu_res = make_sycl_iterator_buffer<element_t>(static_cast<index_t>(1));
   blas::internal::_dot(ex, _N, _vx, _incx, _vy, _incy, gpu_res);
   ex.get_policy_handler().copy_to_host(gpu_res, res.data(), 1);
-  return res[0];
+  return res[0]; //TODO Fix this function and add test
```
Comment (t4c1) [inline src/interface/blas1_interface.hpp:363]: Is this todo something you intended to resolve in this PR?

Comment (fabiomestre) [reply src/interface/blas1_interface.hpp:363]: This is an existing function that I think is bugged (but need to add a test to check). I intended to solve this in this Ticket (after doing the other level one operators) but not on this PR. I forgot to remove the TODO when creating the PR.

File: test/unittest/reduction/reduction_test.cpp [hunk 8]
```diff
@@ -214,4 +214,4 @@ void run_test(const combination_t<scalar_t> combi) {
   ASSERT_TRUE(utils::compare_vectors(out_v_gpu, out_v_cpu));
 }
 
-BLAS_REGISTER_TEST(ReductionPartial, combination_t, combi, generate_name);
+BLAS_REGISTER_TEST_ALL(ReductionPartial, combination_t, combi, generate_name);
```
[the remark this answers is not in the document]
Comment (fabiomestre) [reply test/unittest/reduction/reduction_test.cpp:217]: It was just a refactor to make it more clear what BLAS_REGISTER_TEST is doing. BLAS_REGISTER_TEST is testing all types (float, double, half). But now there is a new function called BLAS_REGISTER_TEST_FLOAT to test only float. So I thought BLAS_REGISTER_TEST_ALL is more meaningful.

File: test/unittest/blas1/blas1_sdsdot_test.cpp (hunk 6 above)
Comment (fabiomestre) [reply test/unittest/blas1/blas1_sdsdot_test.cpp:63]: Yes, N can be zero. However,  the size of the vectors cannot be zero. So,

vectorsSize > 0 and N == 0 // Works
vectorSize == 0 and N >= 0 // Does not work

The error occurs when creating the sycl buffer

File: test/unittest/blas1/blas1_sdsdot_test.cpp (hunk 5 above)
Comment (fabiomestre) [reply test/unittest/blas1/blas1_sdsdot_test.cpp:44]: Fair enough. Will update it

File: src/interface/blas1_interface.hpp (hunk 1 above)
Comment (fabiomestre) [reply src/interface/blas1_interface.hpp:393]: I agree. Will update

File: test/unittest/blas1/blas1_sdsdot_test.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ *  @filename blas1_sdsdot_test.cpp
+ *
+ **************************************************************************/
+
+#include "blas_test.hpp"
+#include <iostream>
+#include <type_traits>
+
+enum class api_type : int { event = 0, result = 1 };
```
[the remark this answers is not in the document]
Comment (Rbiessy) [reply test/unittest/blas1/blas1_sdsdot_test.cpp:30]: I created the ticket

File: include/interface/blas1_interface.h (hunk 2 above)
Comment (fabiomestre) [reply include/interface/blas1_interface.h:81]: Yes, I think so. To be honest I just got the function description from _dot() since it's essentially the same function. Happy to update both descriptions if you think it's more clear

File: include/interface/blas1_interface.h (hunk 3 above)
Comment (fabiomestre) [reply include/interface/blas1_interface.h:94]: Same as above comment. I copied this from _dot() description. Happy to update since it's more clear to refer to them as strides

File: src/interface/blas1_interface.hpp [hunk 10]
```diff
@@ -32,6 +32,7 @@
 #include <vector>
 
 #include "blas_meta.h"
+#include "common/quantization.hpp"
```
[the remark this answers is not in the document]
Comment (fabiomestre) [reply src/interface/blas1_interface.hpp:35]: That was a mistake. Will remove

File: src/interface/blas1_interface.hpp (hunk 1 above)
Comment (fabiomestre) [reply src/interface/blas1_interface.hpp]: Done

File: include/interface/blas1_interface.h (hunk 2 above)
Comment (fabiomestre) [reply include/interface/blas1_interface.h]: Done

File: test/unittest/blas1/blas1_sdsdot_test.cpp (hunk 4 above)
Comment (fabiomestre) [reply test/unittest/blas1/blas1_sdsdot_test.cpp]: Done

File: test/unittest/blas1/blas1_sdsdot_test.cpp (hunk 5 above)
Comment (fabiomestre) [reply test/unittest/blas1/blas1_sdsdot_test.cpp]: Done

File: src/interface/blas1_interface.hpp (hunk 7 above)
Comment (fabiomestre) [reply src/interface/blas1_interface.hpp]: Done

File: src/interface/blas1_interface.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (fabiomestre) [reply src/interface/blas1_interface.hpp]: Done

File: test/unittest/blas1/blas1_sdsdot_test.cpp (hunk 5 above)
Comment (hjabird) [reply test/unittest/blas1/blas1_sdsdot_test.cpp]: Does the test now compile / pass with DPC++? It is currently ComputeCpp only in the unittest cmake.

**Review by Rbiessy — APPROVED** [review]

File: test/unittest/blas1/blas1_sdsdot_test.cpp (hunk 5 above)
Comment (fabiomestre) [reply test/unittest/blas1/blas1_sdsdot_test.cpp]: yes, updated it again to enable the test on dpcpp

**Review by hjabird — APPROVED** [review]
Nice work Fabio.

**Review by KumudhaN — COMMENTED** [review]
Did you get a chance to measure the performance of this kernel against clblast? I think its worth knowing the performance before merging the PR.

File: benchmark/syclblas/blas1/sdsdot.cpp [hunk 11]
```diff
@@ -0,0 +1,128 @@
+/**************************************************************************
+ *
+ *  @license
+ *  Copyright (C) 2016 Codeplay Software Limited
```
Comment (hjabird) [inline benchmark/syclblas/blas1/sdsdot.cpp:4]:
```suggestion
 *  Copyright (C) Codeplay Software Limited
```

File: benchmark/syclblas/blas1/sdsdot.cpp [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+    blas_benchmark::utils::update_counters(state, times);
+  }
+
+  blas_benchmark::utils::calc_avg_counters(state);
+}
+
+template <typename scalar_t>
+void register_benchmark(blas_benchmark::Args& args, ExecutorType* exPtr,
+                        bool* success) {
+  auto gemm_params = blas_benchmark::utils::get_blas1_params(args);
```
Comment (hjabird) [inline benchmark/syclblas/blas1/sdsdot.cpp:111]:
```suggestion
  auto sdsdot_params = blas_benchmark::utils::get_blas1_params(args);
```

**Review by KumudhaN — APPROVED** [review]

File: include/interface/blas1_interface.h [hunk 13]
```diff
@@ -63,19 +63,55 @@ typename executor_t::policy_t::event_t _copy(executor_t &ex, index_t _N,
...
+ * @tparam container_0_t Buffer Iterator
+ * @tparam container_1_t Buffer Iterator
+ * @tparam container_2_t Buffer Iterator
+ * @tparam index_t Index type
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes. If size 0, the result will be sb.
+ * @param sb scalar to add to the results of the inner product.
+ * @param _vx Buffer holding input vector x
+ * @param _incx Stride of vector x
```
Comment (hjabird) [inline include/interface/blas1_interface.h:103]: In bytes or sizeof(data_t)?

File: include/interface/blas1_interface.h [hunk 14]
```diff
@@ -188,6 +232,34 @@ typename ValueType<container_0_t>::type _dot(executor_t &ex, index_t _N,
...
+ * returns the result directly)
+ * @tparam executor_t Executor type
+ * @tparam container_0_t Buffer Iterator
+ * @tparam container_1_t Buffer Iterator
+ * @tparam container_2_t Buffer Iterator
+ * @tparam index_t Index type
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes. If size 0, the result will be sb.
+ * @param sb scalar to add to the results of the inner product.
```
Comment (hjabird) [inline include/interface/blas1_interface.h:248]:
```suggestion
 * @param sb Scalar to add to the results of the inner product.
```

File: include/interface/blas1_interface.h [hunk 15]
```diff
@@ -262,13 +334,22 @@ typename executor_t::policy_t::event_t _copy(executor_t &ex, index_t _N,
...
- * @param _incx Increment for the vector X
- * @param _vx BufferIterator
- * @param _incy Increment for the vector Y
+ * @param _N Input buffer sizes.
+ * @param _vx Buffer holding input vector x
+ * @param _incx Stride of vector x
+ * @param _vy Buffer holding input vector y
+ * @param _incy Stride of vector y
+ * @param _rs output buffer
+ * @return vector of events to wait for.
```
Comment (hjabird) [inline include/interface/blas1_interface.h:352]: Capitalise "output" and "vector" for consistency.

File: include/interface/blas1_interface.h [hunk 16]
```diff
@@ -280,6 +361,36 @@ typename executor_t::policy_t::event_t _dot(
...
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes. If size 0, the result will be sb.
+ * @param sb scalar to add to the results of the inner product.
+ * @param _vx Buffer holding input vector x
+ * @param _incx Stride of vector x
+ * @param _vy Buffer holding input vector y
+ * @param _incy Stride of vector y
+ * @param _rs output buffer
+ * @return vector of events to wait for.
```
Comment (hjabird) [inline include/interface/blas1_interface.h:382]:
```suggestion
 * @param sb Scalar to add to the results of the inner product.
 * @param _vx Buffer holding input vector x
 * @param _incx Stride of vector x
 * @param _vy Buffer holding input vector y
 * @param _incy Stride of vector y
 * @param _rs Output buffer
 * @return Vector of events to wait for.
```

File: include/interface/blas1_interface.h [hunk 17]
```diff
@@ -403,14 +514,22 @@ typename executor_t::policy_t::event_t _rot(
...
- * @param _incx Increment for the vector X
- * @param _vx BufferIterator
- * @param _incy Increment for the vector Y
+ * @param _N Input buffer sizes.
+ * @param _vx Buffer holding input vector x
+ * @param _incx Stride of vector x
+ * @param _vy Buffer holding input vector y
+ * @param _incy Stride of vector y
+ * @param _rs output buffer
+ * @return vector of events to wait for.
```
Comment (hjabird) [inline include/interface/blas1_interface.h:532]:
```suggestion
 * @param _rs Output buffer
 * @return Vector of events to wait for.
```

File: include/interface/blas1_interface.h [hunk 18]
```diff
@@ -423,6 +542,37 @@ typename ValueType<container_0_t>::type _dot(executor_t &ex, index_t _N,
...
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes. If size 0, the result will be sb.
+ * @param sb scalar to add to the results of the inner product.
+ * @param _vx Buffer holding input vector x
+ * @param _incx Stride of vector x
+ * @param _vy Buffer holding input vector y
+ * @param _incy Stride of vector y
+ * @param _rs output buffer
+ * @return vector of events to wait for.
```
Comment (hjabird) [inline include/interface/blas1_interface.h:563]:
```suggestion
 * @param sb Scalar to add to the results of the inner product.
 * @param _vx Buffer holding input vector x
 * @param _incx Stride of vector x
 * @param _vy Buffer holding input vector y
 * @param _incy Stride of vector y
 * @param _rs Output buffer
 * @return Vector of events to wait for.
```

File: src/interface/blas1/dot.cpp.in [hunk 19]
```diff
@@ -35,13 +35,22 @@ namespace blas {
...
+ * @tparam index_t Index type
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes.
+ * @param _vx Buffer holding input vector x
+ * @param _incx Stride of vector x
+ * @param _vy Buffer holding input vector y
+ * @param _incy Stride of vector y
+ * @param _rs output buffer
+ * @return vector of events to wait for.
```
Comment (hjabird) [inline src/interface/blas1/dot.cpp.in:53]:
```suggestion
 * @param _rs Output buffer
 * @return Vector of events to wait for.
```

File: src/interface/blas1/dot_return.cpp.in [hunk 20]
```diff
@@ -35,14 +35,22 @@ namespace blas {
...
+ * @tparam index_t Index type
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes.
+ * @param _vx Buffer holding input vector x
+ * @param _incx Stride of vector x
+ * @param _vy Buffer holding input vector y
+ * @param _incy Stride of vector y
+ * @param _rs output buffer
+ * @return vector of events to wait for.
```
Comment (hjabird) [inline src/interface/blas1/dot_return.cpp.in:53]:
```suggestion
 * @param _rs Output buffer
 * @return Vector of events to wait for.
```

File: src/interface/blas1/sdsdot.cpp.in [hunk 21]
```diff
@@ -0,0 +1,62 @@
...
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes. If size 0, the result will be sb.
+ * @param sb scalar to add to the results of the inner product.
+ * @param _vx Buffer holding input vector x
+ * @param _incx stride of vector x
+ * @param _vy Buffer holding input vector y
+ * @param _incy stride of vector y
+ * @param _rs output buffer
+ * @return vector of events to wait for.
```
Comment (hjabird) [inline src/interface/blas1/sdsdot.cpp.in:55]:
```suggestion
 * @tparam increment_t Increment type
 * @param ex Executor
 * @param _N Input buffer sizes. If size 0, the result will be sb.
 * @param sb Scalar to add to the results of the inner product.
 * @param _vx Buffer holding input vector x
 * @param _incx Stride of vector x
 * @param _vy Buffer holding input vector y
 * @param _incy Stride of vector y
 * @param _rs Output buffer
 * @return Vector of events to wait for.
```

File: src/interface/blas1/sdsdot_return.cpp.in [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename dot_return.cpp.in
```
Comment (hjabird) [inline src/interface/blas1/sdsdot_return.cpp.in:22]:
```suggestion
 *  @filename sdsdot_return.cpp.in
```

File: src/interface/blas1/sdsdot.cpp.in [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename dot.cpp.in
```
Comment (hjabird) [inline src/interface/blas1/sdsdot.cpp.in:22]:
```suggestion
 *  @filename sdsdot.cpp.in
```

File: src/interface/blas1/sdsdot_return.cpp.in [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes. If size 0, the result will be sb.
+ * @param sb scalar to add to the results of the inner product.
+ * @param _vx Buffer holding input vector x
+ * @param _incx stride of vector x
+ * @param _vy Buffer holding input vector y
+ * @param _incy stride of vector y
+ * @param _rs output buffer
+ * @return vector of events to wait for.
```
Comment (hjabird) [inline src/interface/blas1/sdsdot_return.cpp.in:55]:
```suggestion
 * @param _N Input buffer sizes. If size 0, the result will be sb.
 * @param sb Scalar to add to the results of the inner product.
 * @param _vx Buffer holding input vector x
 * @param _incx Stride of vector x
 * @param _vy Buffer holding input vector y
 * @param _incy Stride of vector y
 * @param _rs Output buffer
 * @return Vector of events to wait for.
```

File: src/interface/blas1_interface.hpp [hunk 25]
```diff
@@ -91,13 +91,22 @@ typename executor_t::policy_t::event_t _copy(executor_t &ex, index_t _N,
...
+ * @tparam container_2_t Buffer Iterator
+ * @tparam index_t Index type
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes.
+ * @param _vx Buffer holding input vector x
+ * @param _incx Stride of vector x
+ * @param _vy Buffer holding input vector y
+ * @param _incy Stride of vector y
+ * @param _rs output buffer
```
Comment (hjabird) [inline src/interface/blas1_interface.hpp:108]: Usual capitalisation comment

File: src/interface/blas1_interface.hpp [hunk 26]
```diff
@@ -119,6 +128,43 @@ typename executor_t::policy_t::event_t _dot(
...
+ * @tparam index_t Index type
+ * @tparam increment_t Increment type
+ * @param ex Executor
+ * @param _N Input buffer sizes. If size 0, the result will be sb.
+ * @param sb scalar to add to the results of the inner product.
+ * @param _vx Buffer holding input vector x
+ * @param _incx Stride of vector x
+ * @param _vy Buffer holding input vector y
+ * @param _incy Stride of vector y
+ * @param _rs output buffer
```
Comment (hjabird) [inline src/interface/blas1_interface.hpp:148]: Usual capitalisation comment

File: src/interface/blas1_interface.hpp [hunk 27]
```diff
@@ -325,6 +379,44 @@ typename ValueType<container_0_t>::type _dot(executor_t &ex, index_t _N,
...
```
Comment (hjabird) [inline src/interface/blas1_interface.hpp:399]: Usual capitalisation comment

File: src/interface/blas1_interface.hpp [hunk 28]
```diff
@@ -325,6 +379,44 @@ typename ValueType<container_0_t>::type _dot(executor_t &ex, index_t _N,
...
+                                                increment_t _incx,
+                                                container_1_t _vy,
+                                                increment_t _incy) {
+  using element_t = typename ValueType<container_0_t>::type;
+  element_t res{};
+  auto gpu_res = make_sycl_iterator_buffer<element_t>(static_cast<index_t>(1));
+  auto event1 =
+      blas::internal::_sdsdot(ex, _N, sb, _vx, _incx, _vy, _incy, gpu_res);
+  ex.get_policy_handler().wait(event1);
+  auto event2 = ex.get_policy_handler().copy_to_host(gpu_res, &res, 1);
```
Comment (hjabird) [inline src/interface/blas1_interface.hpp:415]: Is there a way by which the `copy_to_host` can take event1 as a dependcy argument as you can do with lots of USM functions in SYCL? If it could, the first `.wait()` could be avoided.

File: test/blas_test.hpp [hunk 29]
```diff
@@ -257,6 +257,23 @@ struct dump_arg_helper<cl::sycl::half> {
   }
 };
 
+/**
+ * Return type of the tested api (either asynchronous (event) or
+ * synchronous(result)
```
Comment (hjabird) [inline test/blas_test.hpp:262]:
```suggestion
 * Return type of the tested api (either asynchronous (event) or
 * synchronous(result))
```

**Review by hjabird — COMMENTED** [review]
Nitpicking from me! Nice work.

File: src/interface/blas1_interface.hpp (hunk 28 above)
Comment (fabiomestre) [reply src/interface/blas1_interface.hpp:415]: There is no existing function overload that allows that. And I honestly would rather not mess with it to create a new one on this PR.

**Review by mehdi-goli — APPROVED** [review]

File: src/interface/blas1_interface.hpp (hunk 28 above)
Comment (mehdi-goli) [reply src/interface/blas1_interface.hpp:415]: If it is a buffer technically you don't need to wait on that as SYCL handles the dependency, if it is USM you have to pass the event list as an input so that you can explicitly wait on that.

**Review by t4c1 — APPROVED** [review]
